lib/memo_wise.rb

```diff
@@ -163,25 +237,43 @@ def self.has_block_arg?(method) # rubocop:disable Naming/PredicateName
   def self.prepended(target) # rubocop:disable Metrics/PerceivedComplexity
     class << target
       # NOTE: See YARD docs for {.memo_wise} directly below this method!
-      def memo_wise(method_name) # rubocop:disable Metrics/PerceivedComplexity
+      def memo_wise(method_name_or_hash) # rubocop:disable Metrics/PerceivedComplexity
+        klass = self
+        case method_name_or_hash
+        when Symbol
+          method_name = method_name_or_hash
+
+          if klass.singleton_class?
+            MemoWise.create_memo_wise_state!(
+              MemoWise.original_class_from_singleton(klass)
+            )
+          end
+        when Hash
+          unless method_name_or_hash.keys == [:self]
+            raise ArgumentError,
+                  "`:self` is the only key allowed in memo_wise"
+          end
+
+          method_name = method_name_or_hash[:self]
+
+          # In Ruby, "class methods" are implemented as normal instance methods
+          # on the "singleton class" of a given Class object, found via
+          # {Class#singleton_class}.
+          # See: https://medium.com/@leo_hetsch/demystifying-singleton-classes-in-ruby-caf3fa4c9d91
+          MemoWise.create_memo_wise_state!(self)
+          klass = klass.singleton_class
```

Comment:
Can we move this line up under the comment? The comment refers to this line `klass.singleton_class`, rather than the create state line?